In [1]:
# Files from task1_data folder 

import pandas as pd 

# Given features from specifications 
additional_features_df = pd.read_csv('task1_data/additional_features.csv')
# Colour histograms
color_histogram_df = pd.read_csv('task1_data/color_histogram.csv')
# HOG features reduced by PCA
hog_pca_df = pd.read_csv('task1_data/hog_pca.csv')
# Unchanged Test data
test_metadata_df = pd.read_csv('task1_data/test_metadata.csv')
# Unchanged Training data 
train_metadata_df = pd.read_csv('task1_data/train_metadata.csv')

**Experiment 1.1: Basic SVM**

In [2]:
# Naive SVM just to understand what to do

# In this example we will only merge with one additional feature just to get this working 

# In later iterations we will figure out how to improve it

# No Hyperparameter tuning 


from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score


# Associate features with correct labels for training purposes 
additional_features_merged = pd.merge(train_metadata_df, additional_features_df, on="image_id")


# stratify=stratify=train_metadata_df['class_name'] ensures no class imbalance when splitting the data 
# random_state=42 ensures that output is consistent per run
# Splitting the data into a training (which will be further split) and a validation set (which will not be touched)
# local_test_set refers to the test I will be testing for accuracy and other statistics 
training_temp, local_test_set = train_test_split(additional_features_merged, test_size=0.2, random_state=42, stratify=additional_features_merged['class_name'])

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
print(x_trained.columns)
y_trained = training_temp["class_name"]
svm_model = svm.SVC()
svm_model.fit(x_trained, y_trained)
x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_test = local_test_set["class_name"]
predictions = svm_model.predict(x_test)
accuracy = accuracy_score(predictions, y_test)
print(accuracy)



Index(['feat_0', 'feat_1', 'feat_2', 'feat_3', 'feat_4', 'feat_5', 'feat_6',
       'feat_7', 'feat_8', 'feat_9', 'feat_10', 'feat_11', 'feat_12',
       'feat_13', 'feat_14', 'feat_15', 'feat_16', 'feat_17', 'feat_18',
       'feat_19', 'feat_20', 'feat_21', 'feat_22'],
      dtype='object')
0.27066666666666667


In [3]:
# Naive SVM just to understand what to do

# In this example we will only merge with one additional feature just to get this working 

# In later iterations we will figure out how to improve it

# No Hyperparameter tuning 


from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score


# Associate features with correct labels for training purposes 
additional_features_merged = pd.merge(train_metadata_df, hog_pca_df, on="image_id")


# stratify=stratify=train_metadata_df['class_name'] ensures no class imbalance when splitting the data 
# random_state=42 ensures that output is consistent per run
# Splitting the data into a training (which will be further split) and a validation set (which will not be touched)
# local_test_set refers to the test I will be testing for accuracy and other statistics 
training_temp, local_test_set = train_test_split(additional_features_merged, test_size=0.2, random_state=42, stratify=additional_features_merged['class_name'])

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_trained = training_temp["class_name"]
svm_model = svm.SVC()
svm_model.fit(x_trained, y_trained)
x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_test = local_test_set["class_name"]
predictions = svm_model.predict(x_test)
accuracy = accuracy_score(predictions, y_test)
print(accuracy)



0.5026666666666667


In [4]:
# Naive SVM just to understand what to do

# In this example we will only merge with one additional feature just to get this working 

# In later iterations we will figure out how to improve it

# No Hyperparameter tuning 


from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score


# Associate features with correct labels for training purposes 
additional_features_merged = pd.merge(train_metadata_df, color_histogram_df, on="image_id")


# stratify=stratify=train_metadata_df['class_name'] ensures no class imbalance when splitting the data 
# random_state=42 ensures that output is consistent per run
# Splitting the data into a training (which will be further split) and a validation set (which will not be touched)
# local_test_set refers to the test I will be testing for accuracy and other statistics 
training_temp, local_test_set = train_test_split(additional_features_merged, test_size=0.2, random_state=42, stratify=additional_features_merged['class_name'])

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_trained = training_temp["class_name"]
svm_model = svm.SVC()
svm_model.fit(x_trained, y_trained)
x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_test = local_test_set["class_name"]
predictions = svm_model.predict(x_test)
accuracy = accuracy_score(predictions, y_test)
print(accuracy)



0.2613333333333333


<u>Conclusion</u>

The idea of the above experiment was just to see how to actual build a very basic model 

As we can see we did zero hyperparameter testing and no feature selection, leading to low accuracy that is below the threshold 

This suggests that some hyperparameter tuning may potentially improve but we'll see

Accuracy = 0.27066666666666667


**Experiment 1.2 : SVM with HyperParameter Testing**

In [51]:
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV 
from sklearn.svm import SVC 




# Associate features with correct labels for training purposes 
additional_features_merged = pd.merge(train_metadata_df, hog_pca_df, on="image_id")

# stratify=stratify=train_metadata_df['class_name'] ensures no class imbalance when splitting the data 
# random_state=42 ensures that output is consistent per run
# Splitting the data into a training (which will be further split) and a validation set (which will not be touched)
# local_test_set refers to the test I will be testing for accuracy and other statistics 
training_temp, local_test_set = train_test_split(additional_features_merged, test_size=0.2, random_state=42, stratify=additional_features_merged['class_name'])

# Splitting training set further 
# 70% Training, 10% Validation, 20% Test
actual_training_set, actual_validation_set = train_test_split(training_temp, test_size=0.1, random_state=42, stratify=training_temp['class_name'])

# From https://www.geeksforgeeks.org/machine-learning/svm-hyperparameter-tuning-using-gridsearchcv-ml/
hyperparameters = {'C': [0.1, 1, 10, 100, 1000], 
			'gamma': [1, 0.1, 0.01, 0.001, 0.0001], 
			'kernel': ['rbf']}
# Hyperparameter tuning 
hyper_grid = GridSearchCV(SVC(), hyperparameters, refit = True) 
hyper_grid.fit(training_temp.iloc[:, 4:27], training_temp.iloc[:, 3])
predictions = hyper_grid.predict(local_test_set.iloc[:, 4:27])
accuracy = accuracy_score(predictions, local_test_set.iloc[:, 3])
print(accuracy)


# This attempt had an accuracy of 0.476 which is fairly close 
# Difference is I aggregrated with hog pca 






0.476


**Experiment 1.3 : SVM with PCA applied to additional_features**

In [79]:
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

additional_features_merged = pd.merge(train_metadata_df, additional_features_df, on="image_id")

# Split
training_temp, local_test_set = train_test_split(
    additional_features_merged,
    test_size=0.2,
    random_state=42,
    stratify=additional_features_merged['class_name']
)

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_trained = training_temp["class_name"]

x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_test = local_test_set["class_name"]

pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('pca', PCA()),
    ('svc', svm.SVC())
])

hyperparameters = {
    'pca__n_components': [5, 10, 15, 20],
    'svc__C': [0.1, 1, 10, 100],
    'svc__gamma': [1, 0.1, 0.01, 0.001],
    'svc__kernel': ['rbf']
}

# Grid search ONLY on training data
hyper_grid = GridSearchCV(pipeline, hyperparameters, refit=True)
hyper_grid.fit(x_trained, y_trained)

# Predict
predictions = hyper_grid.predict(x_test)

# Evaluate
accuracy = accuracy_score(y_test, predictions)
print(accuracy)

0.48533333333333334


In [80]:
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

# Merge features (keep ONE dataset at a time for fair comparison)
additional_features_merged = pd.merge(train_metadata_df, hog_pca_df, on="image_id")

# Split
training_temp, local_test_set = train_test_split(
    additional_features_merged,
    test_size=0.2,
    random_state=42,
    stratify=additional_features_merged['class_name']
)

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_trained = training_temp["class_name"]

x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_test = local_test_set["class_name"]

# Pipeline (this is now properly used)
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('pca', PCA()),
    ('svc', svm.SVC())
])

# Hyperparameters (clean + aligned with pipeline)
hyperparameters = {
    'pca__n_components': [5, 10, 15, 20],
    'svc__C': [0.1, 1, 10, 100],
    'svc__gamma': [1, 0.1, 0.01, 0.001],
    'svc__kernel': ['rbf']
}

# Grid search ONLY on training data
hyper_grid = GridSearchCV(pipeline, hyperparameters, refit=True)
hyper_grid.fit(x_trained, y_trained)

# Predict
predictions = hyper_grid.predict(x_test)

# Evaluate
accuracy = accuracy_score(y_test, predictions)
print(accuracy)

0.19866666666666666


In [81]:
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

# Merge features (keep ONE dataset at a time for fair comparison)
additional_features_merged = pd.merge(train_metadata_df, color_histogram_df, on="image_id")

# Split
training_temp, local_test_set = train_test_split(
    additional_features_merged,
    test_size=0.2,
    random_state=42,
    stratify=additional_features_merged['class_name']
)

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_trained = training_temp["class_name"]

x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_test = local_test_set["class_name"]

# Pipeline (this is now properly used)
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('pca', PCA()),
    ('svc', svm.SVC())
])

# Hyperparameters (clean + aligned with pipeline)
hyperparameters = {
    'pca__n_components': [5, 10, 15, 20],
    'svc__C': [0.1, 1, 10, 100],
    'svc__gamma': [1, 0.1, 0.01, 0.001],
    'svc__kernel': ['rbf']
}

# Grid search ONLY on training data
hyper_grid = GridSearchCV(pipeline, hyperparameters, refit=True)
hyper_grid.fit(x_trained, y_trained)

# Predict
predictions = hyper_grid.predict(x_test)

# Evaluate
accuracy = accuracy_score(y_test, predictions)
print(accuracy)

0.24133333333333334


**Experiment 1.3 : Deep Learning**

In [86]:
!pip install torch
!pip install torchvision


python(42442) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.


python(42443) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.


In [106]:
# https://www.youtube.com/watch?v=7K-gnWlgAt8
# Extracting features from RESNET - doing own feature engineering 
import torch
from torchvision import models, transforms
import numpy
from PIL import Image
import os

model = models.resnet50(weights=models.ResNet50_Weights.DEFAULT)
model.eval()


transform_pipeline = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    )  
])

feature_extractor = torch.nn.Sequential(*list(model.children())[:-1])
BASE_DIR = os.getcwd() + "/task1_data/"

def extract_features(image_path):
    image = Image.open(image_path).convert("RGB")
    image = transform_pipeline(image)
    image = image.unsqueeze(0)
    with torch.no_grad():
        features = feature_extractor(image)
    features = features.detach().numpy().flatten()
    return features 

features_list = []
for idx, row in train_metadata_df.iterrows():
    image_path = os.path.join(BASE_DIR, row["image_path"])
    feats = extract_features(image_path)
    features_list.append(feats)

features_df = pd.DataFrame(features_list, index=train_metadata_df["image_id"])




In [108]:
# Naive SVM just to understand what to do

# In this example we will only merge with one additional feature just to get this working 

# In later iterations we will figure out how to improve it

# No Hyperparameter tuning 


from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler


# Associate features with correct labels for training purposes 
additional_features_merged = pd.merge(train_metadata_df, features_df, on="image_id")


# stratify=stratify=train_metadata_df['class_name'] ensures no class imbalance when splitting the data 
# random_state=42 ensures that output is consistent per run
# Splitting the data into a training (which will be further split) and a validation set (which will not be touched)
# local_test_set refers to the test I will be testing for accuracy and other statistics 
training_temp, local_test_set = train_test_split(additional_features_merged, test_size=0.2, random_state=42, stratify=additional_features_merged['class_name'])

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
print(x_trained.columns)
y_trained = training_temp["class_name"]
svm_model = svm.SVC()
svm_model.fit(x_trained, y_trained)

x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_test = local_test_set["class_name"]
predictions = svm_model.predict(x_test)
accuracy = accuracy_score(predictions, y_test)
print(accuracy)



Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
       ...
       2038, 2039, 2040, 2041, 2042, 2043, 2044, 2045, 2046, 2047],
      dtype='object', length=2048)
0.9106666666666666


In [ ]:
# Naive SVM just to understand what to do

# In this example we will only merge with one additional feature just to get this working 

# In later iterations we will figure out how to improve it

# No Hyperparameter tuning 


from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler


# Associate features with correct labels for training purposes 
additional_features_merged = pd.merge(train_metadata_df, features_df, on="image_id")


# stratify=stratify=train_metadata_df['class_name'] ensures no class imbalance when splitting the data 
# random_state=42 ensures that output is consistent per run
# Splitting the data into a training (which will be further split) and a validation set (which will not be touched)
# local_test_set refers to the test I will be testing for accuracy and other statistics 
training_temp, local_test_set = train_test_split(additional_features_merged, test_size=0.2, random_state=42, stratify=additional_features_merged['class_name'])

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
print(x_trained.columns)
y_trained = training_temp["class_name"]
svm_model = svm.SVC()
svm_model.fit(x_trained, y_trained)



    
x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_test = local_test_set["class_name"]
predictions = svm_model.predict(x_test)
accuracy = accuracy_score(predictions, y_test)
print(accuracy)



In [ ]:
# Naive SVM just to understand what to do

# In this example we will only merge with one additional feature just to get this working 

# In later iterations we will figure out how to improve it

# No Hyperparameter tuning 


from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler


# Associate features with correct labels for training purposes 
additional_features_merged = pd.merge(train_metadata_df, features_df, on="image_id")


# stratify=stratify=train_metadata_df['class_name'] ensures no class imbalance when splitting the data 
# random_state=42 ensures that output is consistent per run
# Splitting the data into a training (which will be further split) and a validation set (which will not be touched)
# local_test_set refers to the test I will be testing for accuracy and other statistics 
training_temp, local_test_set = train_test_split(additional_features_merged, test_size=0.2, random_state=42, stratify=additional_features_merged['class_name'])
# scaler = StandardScaler()
x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
# x_trained = scaler.fit_transform(x_trained)
y_trained = training_temp["class_name"]
svm_model = svm.SVC()
svm_model.fit(x_trained, y_trained)
x_test = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
# x_test = scaler.transform(x_test)
y_test = local_test_set["class_name"]
predictions = svm_model.predict(x_test)
accuracy = accuracy_score(predictions, y_test)
print(accuracy)





In [ ]:
from sklearn.linear_model import LogisticRegression

combined_features_merged = pd.merge(train_metadata_df, hog_pca_df, on="image_id", how="inner")
combined_features_merged = pd.merge(combined_features_merged, additional_features_df, on="image_id", how="inner")

combined_training_temp, combined_test_set = train_test_split(combined_features_merged, test_size=0.2, random_state=42, stratify=combined_features_merged["class_id"])
combined_x_trained = combined_training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
combined_y_trained = combined_training_temp["class_id"]


pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('logreg', LogisticRegression(max_iter=1000))
    
])

param_grid = {
    'logreg__C': [0.01, 0.1, 1, 10],
    'logreg__solver': ['lbfgs']
}

# Doing cross validation
combined_grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy')
# For hyperparamaeter tuning
combined_grid.fit(combined_x_trained , combined_y_trained)
print(combined_grid.best_params_)

from sklearn.neighbors import KNeighborsClassifier
best_model = combined_grid.best_estimator_

combined_pred = best_model.predict(combined_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"]))
print(f"Accuracy: {accuracy_score(combined_test_set["class_id"], combined_pred)}")


In [ ]:

combined_features_merged = pd.merge(train_metadata_df, hog_pca_df, on="image_id", how="inner")
combined_features_merged = pd.merge(combined_features_merged, additional_features_df, on="image_id", how="inner")

combined_training_temp, combined_test_set = train_test_split(combined_features_merged, test_size=0.2, random_state=42, stratify=combined_features_merged["class_id"])
combined_x_trained = combined_training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
combined_y_trained = combined_training_temp["class_id"]


pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('knn', KNeighborsClassifier())
    
])

param_grid = {
    'knn__n_neighbors': [1, 3, 5, 7, 9, 11, 15, 17],
    'knn__metric': ['euclidean', 'manhattan', 'cosine'],
    'knn__weights': ['uniform', 'distance']
}

# Doing cross validation
combined_grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy')
# For hyperparamaeter tuning
combined_grid.fit(combined_x_trained , combined_y_trained)
print(combined_grid.best_params_)

from sklearn.neighbors import KNeighborsClassifier
best_model = combined_grid.best_estimator_

combined_pred = best_model.predict(combined_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"]))
print(f"Accuracy: {accuracy_score(combined_test_set["class_id"], combined_pred)}")




In [3]:
import torch
from torchvision import models, transforms
import numpy as np
import pandas as pd
import os
from PIL import Image

from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

# ----------------------------
# 1. Load pretrained ResNet50
# ----------------------------
model = models.resnet50(weights=models.ResNet50_Weights.DEFAULT)
model.eval()

feature_extractor = torch.nn.Sequential(*list(model.children())[:-1])


transform_pipeline = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    )
])

BASE_DIR = os.getcwd() + "/task1_data/"


def extract_features(image_path):
    image = Image.open(image_path).convert("RGB")
    image = transform_pipeline(image)
    image = image.unsqueeze(0)

    with torch.no_grad():
        features = feature_extractor(image)

    return features.numpy().flatten()

def build_feature_matrix(df):
    features = []
    for path in df["image_path"]:
        full_path = os.path.join(BASE_DIR, path)
        features.append(extract_features(full_path))
    return np.array(features)

X = build_feature_matrix(train_metadata_df)

# ----------------------------
# 5. Merge labels
# ----------------------------
data = pd.merge(train_metadata_df, pd.DataFrame(X), left_index=True, right_index=True)

# ----------------------------
# 6. Train/validation split
# ----------------------------
training_temp, local_test_set = train_test_split(
    data,
    test_size=0.2,
    random_state=42,
    stratify=data["class_id"]
)

x_trained = training_temp.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_trained = training_temp["class_id"]

x_val = local_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
y_val = local_test_set["class_id"]


scaler = StandardScaler()
x_trained = scaler.fit_transform(x_trained)
x_val = scaler.transform(x_val)


svm_model = svm.SVC(C=10, kernel='rbf', gamma='scale')
svm_model.fit(x_trained, y_trained)

val_pred = svm_model.predict(x_val)
print("Validation Accuracy:", accuracy_score(y_val, val_pred))


def extract_dataset_features(image_paths):
    return scaler.transform([
        extract_features(os.path.join(BASE_DIR, p))
        for p in image_paths
    ])

test_features = extract_dataset_features(test_metadata_df["image_path"])

# 
kaggle_predictions_svm = svm_model.predict(test_features)

submission_df = pd.DataFrame({
    "image_id": test_metadata_df["image_id"],
    "class_id": kaggle_predictions_svm
})

submission_df.to_csv("submission_one.csv", index=False)

Validation Accuracy: 0.8666666666666667


In [4]:
from sklearn.metrics import accuracy_score

val_pred = svm_model.predict(x_val)

accuracy = accuracy_score(y_val, val_pred)
print("Validation Accuracy:", accuracy)

Validation Accuracy: 0.8666666666666667


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

combined_features_merged = pd.merge(train_metadata_df, hog_pca_df, on="image_id", how="inner")
combined_features_merged = pd.merge(combined_features_merged, additional_features_df, on="image_id", how="inner")

combined_training_temp, combined_test_set = train_test_split(
    combined_features_merged,
    test_size=0.2,
    random_state=42,
    stratify=combined_features_merged["class_id"]
)

combined_x_trained = combined_training_temp.drop(
    columns=["image_id", "class_name", "image_path", "class_id"]
)

combined_y_trained = combined_training_temp["class_id"]

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(combined_x_trained)
X_test_scaled = scaler.transform(
    combined_test_set.drop(columns=["image_id", "class_name", "image_path", "class_id"])
)

# Sequential Feature selection

base_model = LogisticRegression(max_iter=1000)

sfs = SequentialFeatureSelector(
    base_model,
    n_features_to_select=50,   
    direction='forward',
    cv=5,
    n_jobs=-1
)

sfs.fit(X_train_scaled, combined_y_trained)

X_train_sel = sfs.transform(X_train_scaled)
X_test_sel = sfs.transform(X_test_scaled)

final_model = LogisticRegression(max_iter=1000)
final_model.fit(X_train_sel, combined_y_trained)

pred = final_model.predict(X_test_sel)

print(f"Accuracy: {accuracy_score(combined_test_set['class_id'], pred)}")

In [ ]:
training_temp, validation_set = train_test_split(
    train_metadata_df, 
    test_size=0.2, 
    random_state=42, 
    stratify=train_metadata_df["class_id"]
)

import torch
from torchvision import models
from sklearn.model_selection import StratifiedKFold
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import numpy as np

resnet = models.resnet50(weights=models.ResNet50_Weights.DEFAULT)
resnet.eval()

resnet_feature_extractor = torch.nn.Sequential(*list(resnet.children())[:-1])
from efficientnet_pytorch import EfficientNet

effnet = EfficientNet.from_pretrained("efficientnet-b0")
effnet.eval()
from torchvision import transforms
from PIL import Image

transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    )
])

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

resnet_feature_extractor = resnet_feature_extractor.to(device)
effnet = effnet.to(device)

def get_resnet_features(image):
    image = transform(image).unsqueeze(0).to(device)

    with torch.no_grad():
        features = resnet_feature_extractor(image)

    return features.squeeze().cpu().numpy()

def get_effnet_features(image):
    image = transform(image).unsqueeze(0).to(device)

    with torch.no_grad():
        features = effnet.extract_features(image)
        features = effnet._avg_pooling(features)
        features = torch.flatten(features, 1)

    return features.squeeze().cpu().numpy()
import numpy as np
import os

BASE_DIR = os.getcwd() + "/task2_data/"



features_list = []
labels_list = []
for idx, row in training_temp.iterrows():
    img_path = os.path.join(BASE_DIR, row["image_path"])
    image = Image.open(img_path).convert("RGB")

    resnet_feat = get_resnet_features(image)
    eff_feat = get_effnet_features(image)
  
    

    combined = np.concatenate([resnet_feat, eff_feat])

    features_list.append(combined)
    labels_list.append(row["class_id"])
X = np.array(features_list)
y = np.array(labels_list)


import numpy as np
from sklearn.metrics import accuracy_score

class StackingClassifier():

    def __init__(self, classifiers, metaclassifier):
        self.classifiers = classifiers
        self.metaclassifier = metaclassifier

    def fit(self, X, y):
        # train base models
        for clf in self.classifiers:
            clf.fit(X, y)

        # build meta features
        X_meta = self._predict_base(X)

        # train meta model
        self.metaclassifier.fit(X_meta, y)

    def _predict_base(self, X):
        yhats = []

        for clf in self.classifiers:
            yhat = clf.predict_proba(X)   # [N, C]
            yhats.append(yhat)

        yhats = np.hstack(yhats)  # [N, C * num_models]
        return yhats

    def predict(self, X):
        X_meta = self._predict_base(X)
        return self.metaclassifier.predict(X_meta)

    def score(self, X, y):
        preds = self.predict(X)
        return accuracy_score(y, preds)
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
SVC(probability=True)
base_models = [
    LogisticRegression(max_iter=2000),
    RandomForestClassifier(n_estimators=100),
    SVC(probability=True)
]

meta_model = LogisticRegression(max_iter=2000)



In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from efficientnet_pytorch import EfficientNet


#https://www.codegenes.net/blog/efficientnetpytorch/
!pip install efficientnet-pytorch
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from efficientnet_pytorch import EfficientNet
from PIL import Image
from torchvision import transforms
from torch.utils.data import Dataset, DataLoader
from sklearn.model_selection import train_test_split
import random
import numpy as np


# Making sure results are the same every time it runs
# https://docs.pytorch.org/docs/2.12/notes/randomness.html
SEED_VAR = 42
random.seed(SEED_VAR)
torch.manual_seed(SEED_VAR)
np.random.seed(SEED_VAR)
torch.cuda.manual_seed(SEED_VAR)
torch.cuda.manual_seed_all(SEED_VAR)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
generator_val = torch.Generator()
generator_val.manual_seed(SEED_VAR)

BASE_DIR = os.getcwd() + "/task2_data/"

training_df, val_df = train_test_split(
    train_metadata_df, 
    test_size=0.2, 
    random_state=42, 
    stratify=train_metadata_df["class_id"]
)

training_transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
])

val_transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    )])

# Adapted from https://www.kaggle.com/code/veb101/transfer-learning-using-efficientnet-models
class CustomDataset(Dataset):
    def __init__(self, df, root_dir, transform=None):
        self.df = df
        self.transform = transform
        self.root_dir = root_dir
        
    def __len__(self):
        return len(self.df)
    
    def __getitem__(self, idx):
        row = self.df.iloc[idx]  
        img_path = os.path.join(self.root_dir, str(row["image_path"]))
        label = row["class_id"]
        image = Image.open(img_path).convert("RGB")
        if self.transform:
            image = self.transform(image)
        return image, label

train_dataset = CustomDataset(training_df, BASE_DIR, training_transform)
val_dataset = CustomDataset(val_df, BASE_DIR, val_transform)

train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=32, shuffle=True)
val_loader = torch.utils.data.DataLoader(val_dataset, batch_size=32, shuffle=False)

 
model = EfficientNet.from_pretrained('efficientnet-b0')
num_ftrs = model._fc.in_features
model._fc = nn.Linear(num_ftrs, len(train_metadata_df["class_id"].unique()))
 
criterion = nn.CrossEntropyLoss()
optimizer = optim.AdamW(model.parameters(), lr=1e-4)
 
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
model.to(device)

def evaluate(model, val_loader, criterion, device):
    model.eval()

    val_loss = 0
    correct = 0
    total = 0

    with torch.no_grad():
        for inputs, labels in val_loader:
            inputs = inputs.to(device)
            labels = labels.to(device)

            outputs = model(inputs)
            loss = criterion(outputs, labels)

            val_loss += loss.item()

            _, preds = torch.max(outputs, 1)
            correct += (preds == labels).sum().item()
            total += labels.size(0)

    acc = correct / total
    return val_loss / len(val_loader), acc
 
# for epoch in range(10):
#     model.train()
#     running_loss = 0.0
#     for i, data in enumerate(train_loader, 0):
#         inputs, labels = data[0].to(device), data[1].to(device)
 
#         optimizer.zero_grad()
 
#         outputs = model(inputs)
#         loss = criterion(outputs, labels)
#         loss.backward()
#         optimizer.step()
 
#         running_loss += loss.item()
#     print(f'Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}')
#     val_loss, val_acc = evaluate(best_model, val_loader, criterion, device)

#     print(
#         f"Epoch {epoch+1} | "
#         f"Train Loss: {running_loss/len(train_loader):.4f} | "
#         f"Val Loss: {val_loss:.4f} | "
#         f"Val Acc: {val_acc:.4f}"
#     )
 
import copy

best_val_acc = 0
epoch_val = -1
for epoch in range(21):
    model.train()
    running_loss = 0.0
    for i, data in enumerate(train_loader, 0):
        inputs, labels = data[0].to(device), data[1].to(device)
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
    print(f'Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}')
    val_loss, val_acc = evaluate(model, val_loader, criterion, device)

    print(
        f"Epoch {epoch+1} | "
        f"Train Loss: {running_loss/len(train_loader):.4f} | "
        f"Val Loss: {val_loss:.4f} | "
        f"Val Acc: {val_acc:.4f}"
    )
    # https://debuggercafe.com/saving-and-loading-the-best-model-in-pytorch/
    # we can save the best mode to a pth file 
    if val_acc > best_val_acc:
        torch.save(model.state_dict(), "best_model.pth")
        best_val_acc = val_acc
        epoch_val = epoch








# Loading the model from best epoch
best_model =  EfficientNet.from_pretrained('efficientnet-b0')
num_ftrs = best_model._fc.in_features
best_model._fc = nn.Linear(num_ftrs, len(train_metadata_df["class_id"].unique()))
best_model.load_state_dict(torch.load("best_model.pth", map_location=device))
best_model.to(device)
best_model.eval()

val_loss, val_acc = evaluate(best_model, val_loader, criterion, device)
print(val_acc)
class TestDataset(Dataset):
    def __init__(self, df, root_dir, transform=None):
        self.df = df
        self.transform = transform
        self.root_dir = root_dir
        
    def __len__(self):
        return len(self.df)
    
    def __getitem__(self, idx):
        row = self.df.iloc[idx]  
        img_path = os.path.join(self.root_dir, str(row["image_path"]))
        image = Image.open(img_path).convert("RGB")
        if self.transform:
            image = self.transform(image)
        return image

test_dataset = TestDataset(test_metadata_df, BASE_DIR, val_transform)

test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=32, shuffle=False)

def predict(model, loader, device):
    model.eval()
    preds = []

    with torch.no_grad():
        for images in loader:
            images = images.to(device)

            outputs = model(images)
            _, predicted = torch.max(outputs, 1)

            preds.extend(predicted.cpu().numpy())

    return preds

test_preds = predict(best_model, test_loader, device)

submission_df = pd.DataFrame({
    "image_id": test_metadata_df["image_id"],
    "class_id": test_preds
})

submission_df.to_csv("cnn_task_2.csv", index=False)



In [ ]:
import torch
from torchvision import models
from sklearn.model_selection import StratifiedKFold
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.utils import resample
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import StratifiedKFold



resnet = models.resnet50(weights=models.ResNet50_Weights.DEFAULT)
resnet.eval()

resnet_feature_extractor = torch.nn.Sequential(*list(resnet.children())[:-1])
from efficientnet_pytorch import EfficientNet

effnet = EfficientNet.from_pretrained("efficientnet-b0")
effnet.eval()
from torchvision import transforms
from PIL import Image

transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    )
])

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

resnet_feature_extractor = resnet_feature_extractor.to(device)
effnet = effnet.to(device)

def get_resnet_features(image):
    image = transform(image).unsqueeze(0).to(device)

    with torch.no_grad():
        features = resnet_feature_extractor(image)

    return features.squeeze().cpu().numpy()

def get_effnet_features(image):
    image = transform(image).unsqueeze(0).to(device)

    with torch.no_grad():
        features = effnet.extract_features(image)
        features = effnet._avg_pooling(features)
        features = torch.flatten(features, 1)

    return features.squeeze().cpu().numpy()


import numpy as np
import os

BASE_DIR = os.getcwd() + "/task2_data/"

features_list = []
labels_list = []
for idx, row in train_metadata_df.iterrows():
    img_path = os.path.join(BASE_DIR, row["image_path"])
    image = Image.open(img_path).convert("RGB")

    resnet_feat = get_resnet_features(image)
    eff_feat = get_effnet_features(image)
    combined = np.concatenate([resnet_feat, eff_feat])

    features_list.append(combined)
    labels_list.append(row["class_id"])
X = np.array(features_list)
y = np.array(labels_list)

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

# skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
# def train_bag(X_train, y_train, n_models):
#     models = []
    

    
for i in range(n_models):

    X_bag, y_bag = resample(X_train, y_train, replace=True, random_state=42+i)

    model = LogisticRegression(max_iter=5000, solver="saga")
    model.fit(X_bag, y_bag)

    models.append(model)
def bagging_predict(models, X):
    preds = []

    for model in models:
        preds.append(model.predict_proba(X))  
    preds = np.mean(preds, axis=0)  
    return np.argmax(preds, axis=1)

final_preds = bagging_predict(models, X_test)
print("Bagging accuracy:", accuracy_score(y_test, final_preds))








